src/compartment.c

```diff
@@ -139,192 +142,302 @@ comp_from_elf(char* filename, struct ConfigEntryPoint* entry_points, size_t entr
             {
-                new_comp->relas[j] = comp_relas[j].r_offset;
-                // TODO double check
-                /*comp_relas[j].r_offset += new_comp->base;*/
-                /*uintptr_t old_plt_val = (uintptr_t) *((void**) comp_relas[j].r_offset);*/
-                /*old_plt_val += new_comp->base;*/
+                new_entry_point->fn_addr = ep_syms[i].st_value;
+                break;
             }
+            default:
+                assert(false);
+        }
+        new_comp->comp_fns[new_comp->entry_point_count] = new_entry_point;
+        new_comp->entry_point_count += 1;
+    }
+    free(ep_syms);
 
-            free(comp_relas);
+    // Find symbols for intercepts
+    char** intercept_names = calloc(intercept_count, sizeof(char*));
+    const char* so_plt_suffix = "@plt";
+    for (size_t i = 0; i < intercept_count; ++i)
+    {
+        if (new_comp->elf_type == ET_DYN)
+        {
+            size_t to_intercept_name_len = strlen(intercepts[i]) + strlen(so_plt_suffix) + 1;
+            intercept_names[i] = malloc(to_intercept_name_len);
+            strcpy(intercept_names[i], intercepts[i]);
+            strcat(intercept_names[i], so_plt_suffix);
+        }
+        else if (new_comp->elf_type == ET_EXEC)
+        {
+            intercept_names[i] = malloc(strlen(intercepts[i]) + 1);
+            strcpy(intercept_names[i], intercepts[i]);
         }
         else
+        {
+            assert(false && "Shouldn't get here");
+        }
+    }
+    Elf64_Sym* intercept_syms =
+        find_symbols((const char**) intercept_names, intercept_count, false,
+                     comp_symtb, comp_strtb, comp_symtb_shdr.sh_size);
+    for (size_t i = 0; i < intercept_count; ++i)
+    {
+        // TODO better way to check if we didn't find an intercept?
+        if (intercept_syms[i].st_value != 0)
+        {
+            comp_add_intercept(new_comp, intercept_syms[i].st_value, intercept_addrs[i]);
+        }
+        free(intercept_names[i]);
+    }
+    free(intercept_names);
+    free(intercept_syms);
+
+    // Find all symbols for eager relocation mapping
+    for (size_t i = 0; i < new_comp->rela_maps_count; ++i)
+    {
+        // Ignore relocations we don't want to load, as earlier set on lookup
+        // (e.g., weak-bound symbols)
+        if (new_comp->rela_maps[i].rela_address == 0)
         {
             continue;
         }
+        bool found = false;
+        for (size_t j = 0; j < new_comp->lib_deps_count; ++j)
+        {
+            for (size_t k = 0; k < new_comp->lib_deps[j]->lib_syms_count; ++k)
+            {
+                if (!strcmp(new_comp->rela_maps[i].rela_name,
+                            new_comp->lib_deps[j]->lib_syms[k].sym_name))
+                {
+                    new_comp->rela_maps[i].target_func_address =
+                        new_comp->lib_deps[j]->lib_mem_base +
+                        new_comp->lib_deps[j]->lib_syms[k].sym_offset;
+                    found = true;
```

Comment:
This looks like a good place for using a C `goto`?

Reply:
Done in https://github.com/capablevms/CHERI-ELF-comp/pull/13/commits/cf9abc986807ead41dfefcbb5db926870cc758fc.